mixing/mixpool/mixpool.go

```diff
@@ -0,0 +1,1523 @@
+type entry struct {
+	hash     chainhash.Hash
+	sid      [32]byte
+	recvTime time.Time
+	msg      mixing.Message
+	msgtype  msgtype
+	run      uint32
+}
+
+type orphan struct {
+	message  mixing.Message
+	accepted time.Time
+}
+
+type session struct {
+	sid    [32]byte
+	runs   []runstate
+	expiry uint32
+	bc     broadcast
+}
+
+type runstate struct {
+	run    uint32
+	prs    []chainhash.Hash
+	counts [nmsgtypes]uint32
+	hashes map[chainhash.Hash]struct{}
+}
+
+func (r *runstate) countFor(t msgtype) uint32 {
+	return r.counts[t-1]
+}
+
+func (r *runstate) incrementCountFor(t msgtype) {
+	r.counts[t-1]++
+}
+
+type broadcast struct {
+	ch chan struct{}
+	mu sync.Mutex
+}
+
+// wait returns the wait channel that is closed whenever a message is received
+// for a session.  Waiters must acquire the pool lock before reading messages.
+func (b *broadcast) wait() <-chan struct{} {
+	b.mu.Lock()
+	ch := b.ch
+	b.mu.Unlock()
+
+	return ch
+}
+
+func (b *broadcast) signal() {
+	b.mu.Lock()
+	close(b.ch)
+	b.ch = make(chan struct{})
+	b.mu.Unlock()
+}
+
+// Pool records in-memory mix messages that have been broadcast over the
+// peer-to-peer network.
+type Pool struct {
+	mtx                sync.RWMutex
+	prs                map[chainhash.Hash]*wire.MsgMixPairReq
+	outPoints          map[wire.OutPoint]chainhash.Hash
+	pool               map[chainhash.Hash]entry
+	orphans            map[chainhash.Hash]*orphan
+	orphansByID        map[idPubKey]map[chainhash.Hash]mixing.Message
+	messagesByIdentity map[idPubKey][]chainhash.Hash
+	latestKE           map[idPubKey]*wire.MsgMixKeyExchange
+	sessions           map[[32]byte]*session
+	sessionsByTxHash   map[chainhash.Hash]*session
+	epoch              time.Duration
+	expireHeight       uint32
+	expireSem          chan struct{}
+
+	blockchain  BlockChain
+	utxoFetcher UtxoFetcher
+	feeRate     int64
+	params      *chaincfg.Params
+}
+
+// UtxoEntry provides details regarding unspent transaction outputs.
+type UtxoEntry interface {
+	IsSpent() bool
+	PkScript() []byte
+	ScriptVersion() uint16
+	BlockHeight() int64
+	Amount() int64
+}
+
+// UtxoFetcher defines methods used to validate unspent transaction outputs in
+// the pair request message.  It is optional, but should be implemented by full
+// nodes that have this capability to detect and stop relay of spam and junk
+// messages.
+type UtxoFetcher interface {
+	// FetchUtxoEntry defines the function to use to fetch unspent
+	// transaction output information.
+	FetchUtxoEntry(wire.OutPoint) (UtxoEntry, error)
+}
+
+// BlockChain queries the current status of the blockchain.  Its methods should
+// be able to be implemented by both full nodes and SPV wallets.
+type BlockChain interface {
+	// ChainParams identifies which chain parameters the mixing pool is
+	// associated with.
+	ChainParams() *chaincfg.Params
+
+	// BestHeader returns the hash and height of the current tip block.
+	BestHeader() (chainhash.Hash, int64)
```

Comment:
I would advise against calling this `BestHeader` because it does not have the same meaning as already existing code which uses that name for something else.  As the comment on `internal/blockchain/BlockChain.BestHeader` states:

> // BestHeader returns the header with the most cumulative work that is NOT
> // known to be invalid.

That is **NOT** the same thing as the current tip block that has been validated.

Reply:
To be more explicit the real distinction here is what the current target best tip header is versus what the actual current validated chain tip is.  The former is what `internal/blockchain/BlockChain.BestHeader` returns and the latter is obtained via the `Height` and `Hash` fields of the struct returned by `internal/blockchain/BlockChain.BestSnapshot`.

Reply:
You could probably call this `BestTip` instead and then satisfy the interface when creating the pool via a small shim adaptor like:

```Go
type mixPoolChainAdaptor struct {
	*blockchain.BlockChain
}

func (c *mixPoolChainAdaptor) BestTip() (chainhash.Hash, int64) {
	best := c.BlockChain.BestSnapshot()
	return best.Hash, best.Height
}

...

mixPool := NewPool(&mixPoolChainAdaptor{s.chain})
```

Reply:
yes, it is passing in (through an embedded struct member) blockchain.BlockChain.  will rename.

Reply:
changing to

```
	// CurrentTip returns the hash and height of the current tip block.
	CurrentTip() (chainhash.Hash, int64)
```